crates/service/src/middleware/inject_allocation.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+//! injects allocation id in extensions
+//! - check if allocation id already exists
+//! - else, try to fetch allocation id from deployment_id and allocations watcher
+//! - execute query
+//!
+//! Needs signed receipt Extension to be added OR deployment id
+
+use std::collections::HashMap;
+
+use alloy::primitives::Address;
+use axum::{
+    extract::{Request, State},
+    middleware::Next,
+    response::Response,
+};
+use tap_core::receipt::SignedReceipt;
+use thegraph_core::DeploymentId;
+use tokio::sync::watch;
+
+#[derive(Clone)]
+pub struct Allocation(pub Address);
+
+impl From<Allocation> for String {
+    fn from(value: Allocation) -> Self {
+        value.0.to_string()
+    }
+}
+
+#[derive(Clone)]
+pub struct AllocationState {
+    pub deployment_to_allocation: watch::Receiver<HashMap<DeploymentId, Address>>,
+}
+
+pub async fn allocation_middleware(
+    State(my_state): State<AllocationState>,
+    mut request: Request,
+    next: Next,
+) -> Response {
+    if let Some(receipt) = request.extensions().get::<SignedReceipt>() {
+        let allocation = receipt.message.allocation_id;
+        request.extensions_mut().insert(Allocation(allocation));
+    } else if let Some(deployment_id) = request.extensions().get::<DeploymentId>() {
```

Comment:
This is not related to receipt, every receipt contains allocation id. There are free queries that only need Authorization instead of a receipt, we still want to inject that in the metrics.

Reply:
Chatted offline. We're good.